llvm/lib/IR/ConstantRange.cpp

```diff
@@ -364,6 +364,29 @@ ConstantRange ConstantRange::makeExactNoWrapRegion(Instruction::BinaryOps BinOp,
   return makeGuaranteedNoWrapRegion(BinOp, ConstantRange(Other), NoWrapKind);
 }
 
+ConstantRange ConstantRange::makeMaskNotEqualRange(const APInt &Mask,
+                                                   const APInt &C) {
+  assert(!Mask.isZero() && "Mask cannot be zero.");
+  unsigned BitWidth = Mask.getBitWidth();
+  unsigned TrailingZeroesOfMask = Mask.countr_zero();
+
+  // If (Val & Mask) != 0 then the value must be larger than the lowest set
+  // bit of Mask.
+  if (C.isZero())
+    return ConstantRange::getNonEmpty(
+        APInt::getOneBitSet(BitWidth, TrailingZeroesOfMask),
+        APInt::getZero(BitWidth));
```

Comment:
Do we need this special case? I think for C=0 the below case is equivalent to this one...

Reply:
It it is indeed equivalent, but thought it was a bit nice / helpful to have the base case explicit.

Reply:
IMHO its best not to have unnecessary special cases...